### PR TITLE
Fix coverage in src layout with test.

### LINKED ISSUE #3636: Coverage plugin is broken on Cython modules in packages with sources in src-layout
_(Ref: https://twitter.com/webKnjaZ/status/1265373614003625987)_

Having a module under src-layout causes `Cython.Coverage.Plugin` to not set itself as a tracer in the DB. And even when it's set, it fails to matchthe module with its sources

I've made a reproducer repo with GitHub Actions Workflow integrated as a demo of this bug:
- https://github.com/webknjaz/cython-coverage-extern-bug-repro-issue3636
- https://github.com/webknjaz/cython-coverage-extern-bug-repro-issue3636/runs/718112079?check_suite_focus=true#step:13:6

When this happens, the following message is printed out during the report stage:
```console
Couldn't parse '..../src/call_mymath/call_mymath.pyx' as Python source: 'invalid syntax' at line 1
```

Reproducer details:
==================
It's based on the example https://github.com/cython/cython/tree/master/Demos/libraries with a slightly modified structure:
```console
|- mymath.c
|- mymath.h
|- setup.py
|- src/
    |- call_mymath/
        |- call_mymath.pyx
        |- __init__.py
    |-flat_ns_call_mymath.pyx
|- test_cython.py
```
`flat_ns_call_mymath.pyx` is the same as `call_mymath.pyx` (the only thing modified inside is relative path to the C-header).

`setup.py` puts both `flat_ns_call_mymath.so` and `call_mymath.so` under `call_mymath` package in `site-packages`. So they are importable as `call_mymath.flat_ns_call_mymath` and `call_mymath.call_mymath`. `call_mymath.flat_ns_call_mymath` shows up in the report and `call_mymath.call_mymath` doesn't because `Cython.Coverage.Plugin` fails to map it with the source and claim the file tracing rights.

STR
===
1. `pip install coverage Cython`
2. `CFLAGS=-DCYTHON_TRACE=1 cythonize --inplace --directive=embedsignature=True --directive=emit_code_comments=True --directive=linetrace=True --directive=profile=True '**/*.pyx'`
3. `pip install .`
4. `coverage run -m test_cython`
5. `coverage report` -- this is where it fails with
    ```console
    Couldn't parse '..../src/call_mymath/call_mymath.pyx' as Python source: 'invalid syntax' at line 1
    ```
    (adding `-i` makes at least `flat_ns_call_mymath` show up correctly)

When looking at `.coverage` using `sqlite3`, `flat_ns_call_mymath.pyx` is mapped to `Cython.Coverage.Plugin` in the `tracer` table while there's no record for `call_mymath.pyx` there.

### PR DESCRIPTION
**Pull request by rhpvorderman**
This fixes #3636 
This also obsoletes #3648 . #3648 was taken as a base. A test was added. Also @scoder 's feedback on that PR has been incorporated in this request. 

Thanks @webknjaz for providing the initial fix.

To make maintainer lives easier, I have only used os.path so it should be portable to the 0.29.x series as well that should be python 2.7 compatible. I volunteer to do so if this PR is merged.

### CODE REVIEW

File: Cython/Coverage.py [hunk 1]
```diff
@@ -41,6 +41,21 @@ def _find_dep_file_path(main_file, file_path, relative_path_search=False):
         rel_file_path = os.path.join(os.path.dirname(main_file), file_path)
         if os.path.exists(rel_file_path):
             abs_path = os.path.abspath(rel_file_path)
+
+        abs_no_ext = os.path.splitext(abs_path)[0]
+        file_no_ext, extension = os.path.splitext(file_path)
+        # We check if the paths match by matching the directories in reverse order.
+        # pkg/module.pyx /long/absolute_path/bla/bla/site-packages/pkg/module.c should match.
+        # this will match the pairs: module-module and pkg-pkg. After which there is nothing left to zip.
+        matching_paths = zip(reversed(abs_no_ext.split(os.sep)), reversed(file_no_ext.split(os.sep)))
```
Comment (webknjaz) [inline Cython/Coverage.py:50]: Should this also normalize paths?
```suggestion
        matching_paths = zip(
            reversed(map(os.path.normpath, abs_no_ext.split(os.sep))),
            reversed(map(os.path.normpath, file_no_ext.split(os.sep))),
        )
```

**Comment by webknjaz** [comment]
I've pointed my reproducer repo to this branch and it's still failing for some reason. See: https://github.com/webknjaz/cython-coverage-extern-bug-repro-issue3636/runs/1125597281#step:13:6.

@rhpvorderman mind checking that repro against your fix?

**Comment by rhpvorderman** [comment]
> I've pointed my reproducer repo to this branch and it's still failing for some reason. See: https://github.com/webknjaz/cython-coverage-extern-bug-repro-issue3636/runs/1125597281#step:13:6.
> 
> @rhpvorderman mind checking that repro against your fix?

@webknjaz The coverage plugin checks for a valid `.c` file not a `.pyx` file to check the coverage. The coverage plugin can't work without these C files. The c files have to be somewhere. If you already have them, where are they located? 

In the example I generate the C files in place. That's why I added https://github.com/cython/cython/blob/9311d55f5afcbc3591550811922e85f8df27038d/tests/run/coverage_cmd_src_pkg_layout.srctree#L6 
to the example. A `cythonize -X linetrace=True  path/site-packages/mypkg/*.pyx` should also work.

**Comment by rhpvorderman** [comment]
CI integration fails on a timeout. Seems the fix doesn't break any tests.

File: Cython/Coverage.py (hunk 1 above)
Comment (da-woods) [reply Cython/Coverage.py:50]: I'm also not too keen on `str.split` here - I wonder if `os.path.split` would be more robust (but it'd need multiple calls to completely split a path)

**Review by da-woods — COMMENTED** [review]
Sorry for the "slightly" slow review on this....

File: Cython/Coverage.py (hunk 1 above)
Comment (rhpvorderman) [reply Cython/Coverage.py:52]: os.path.spliit is implemented using common string methods (rfind) that use the separator. So it is not necessarily more robust.

**Comment by rhpvorderman** [comment]
> Sorry for the "slightly" slow review on this....

No problem! I like how Cython has introduced me to the Python C-API. Eventually I started using cython more and more to connect to the C-API and currently all my projects are just written in C directly. So while I still have warm feelings for Cython, I am affraid my personal use case for this PR has vanished. As a result I do not have the means to test it anymore. Maybe @webknjaz has?

**Comment by da-woods** [comment]
Thanks for coming back to this @rhpvorderman. I suspect I can get this into a mergeable state myself from here

**Comment by rhpvorderman** [comment]
@da-woods Thanks again for your help and advice back two years ago: https://github.com/cython/cython/issues/3388#issuecomment-683672261 and https://github.com/cython/cython/issues/3388#issuecomment-683679403.

**Comment by da-woods** [comment]
Thank you @rhpvorderman and @webknjaz

File: tests/run/coverage_cmd_src_pkg_layout.srctree [hunk 2]
```diff
@@ -0,0 +1,177 @@
+# mode: run
+# tag: coverage,trace
+
+"""
+PYTHON -m pip install .
```
Comment (scoder) [inline tests/run/coverage_cmd_src_pkg_layout.srctree:5]: This line seems to be required (the test fails without it), but it modifies the current venv. A test should never do that. See the report in https://github.com/cython/cython/issues/5788

**Pull request merged**